docs/Migration-Guide-V3.md

```diff
@@ -113,45 +129,64 @@ const ajv = new AJV();
 
-From Fastify v3, the behavior of `preParsing` hook will change slightly in order to support request payload manipulation.
+From Fastify v3, the behavior of the `preParsing` hook will change slightly
+in order to support request payload manipulation.
 
-The hook now takes an additional argument, `payload`, and therefore the new hook signature is `fn(request, reply, payload, done)` or `async fn(request, reply, payload)`.
+The hook now takes an additional argument, `payload`, and therefore the new hook
+signature is `fn(request, reply, payload, done)` or
+`async fn(request, reply, payload)`.
 
-The hook can optionally return a new stream via `done(null, stream)` or returning the stream in case of async functions.
+The hook can optionally return a new stream via `done(null, stream)` or
+returning the stream in case of async functions.
 
-If the hook returns a new stream, it will be used instead of the original one in following hooks. A sample use case for this is handling compressed requests.
+If the hook returns a new stream, it will be used instead of the original one in
+following hooks. A sample use case for this is handling compressed requests.
 
-The new stream should add the `receivedEncodedLength` property to the stream that should reflect the actual data size received from the client. For instance, in compressed request it should be the size of the compressed payload. 
+The new stream should add the `receivedEncodedLength` property to the stream
+that should reflect the actual data size received from the client. For instance,
+in a compressed request it should be the size of the compressed payload.
 This property can (and should) be dynamically updated during `data` events.
 
-The old syntax of Fastify v2 without payload it is supported but it is deprecated.
+The old syntax of Fastify v2 without payload is supported but it is deprecated.
 
 ### Changed hooks behaviour ([#2004](https://github.com/fastify/fastify/pull/2004))
 
-From Fastify v3, the behavior of `onRoute` and `onRegister` hooks will change slightly in order to support hook encapsulation.
+From Fastify v3, the behavior of `onRoute` and `onRegister` hooks will change
+slightly in order to support hook encapsulation.
 
-- `onRoute` - The hook will be called asynchronously, in v1/v2 it's called as soon as a route is registered. This means that if you want to use it, you should register this hook as soon as possible in your code.
-- `onRegister` - Same as the onRoute hook, the only difference is that now the very first call will no longer be the framework itself, but the first registered plugin
+- `onRoute` - The hook will be called asynchronously. In v2 this hook is called
+as soon as a route is registered. This means that if you want to use it, you
+should register this hook as soon as possible in your code.
+(TODO: this statement is unclear. Are we saying that the v2 hook should be
+registered as soon as possible? Or are we saying the v3 hook should be?)
```

Comment:
in v2 this hook was run when the code was defining the route. In v3, we are calling the hook after all plugins are registered before ready.

Reply:
I'm still not clear. An example of how things worked in v2 versus v3 would help.

Reply:
Look at the tests of.

Reply:
@jsumners consider https://github.com/fastify/fastify/pull/1971/files#diff-1934f8b6f3b9c6ad263f75fb75f996deR2711-R2738. We had to move `onRegister` _before_ the plugin definition.
The same thing is true for `onRoute`.

Reply:
can you update it and resolve?

Reply:
I'm still just not clear what the text should be. If someone has any suggestions, that'd be great.

Reply:
```
The hook is now inherited when registered a new plugin: you know need to add the `onRoute`  hook _before_ registering any plugins, and not after.
```

Is this clearer?

Reply:
I think I can work with that. I'll deal with it soon.

Reply:
Please re-review.